### ISSUE TITLE
Allow explicit feature edge detection method during meshing

### ISSUE THREAD

**Issue by TheodoreGalanos** [ladybug-analysis-tools/butterfly#165]
Hi I think it's time to add this one.

It's not hard to do, the steps required are:

1. Running a feature edge detection on the .stl file in question. The user should be able to select which file to run it on, thus I feel this options should be integrated into the BF_geometry component. The command to run (right before SHM) is simply `surfaceFeatureExtract`. The options are included in a surfaceFeatureExtractDict which looks like this and should be placed in the system directory:

```
/*--------------------------------*- C++ -*----------------------------------*\
| =========                 |                                                 |
| \\      /  F ield         | OpenFOAM: The Open Source CFD Toolbox           |
|  \\    /   O peration     | Version:  4.x                                   |
|   \\  /    A nd           | Web:      www.OpenFOAM.org                      |
|    \\/     M anipulation  |                                                 |
\*---------------------------------------------------------------------------*/
FoamFile
{
    version     2.0;
    format      ascii;
    class       dictionary;
    object      surfaceFeatureExtractDict;
}
// * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * //


surface1.stl
{
    // How to obtain raw features (extractFromFile || extractFromSurface)
    extractionMethod    extractFromSurface;

    extractFromSurfaceCoeffs
    {
        // Mark edges whose adjacent surface normals are at an angle less
        // than includedAngle as features
        // - 0  : selects no edges
        // - 180: selects all edges
        includedAngle   120;

        // Do not mark region edges
        geometricTestOnly       yes;
    }

    // Write options

        // Write features to obj format for postprocessing
        writeObj                yes;
}
```

The other parameter the user can select is the includedAngle. The default seems to be ok, but users can also experiment. The value range is written above.

2. The above command will create an .eMesh file in the triSurface directory which is then included in the SnappyHexMeshDict in the `features {}` section like the example below:

`  features
  (
      {
          file ”file_name.eMesh”; // file containing edge mesh
          level 2;               // level of refinement
      }
  );
`

3. Lastly, the implicit and multiregionsnap options in SHMDict should be set to false, while the explicit value should be set to true.

That's it! We now have explicit feature detection. The actual refinement happens in the first stages of the SHM process, users can see information about it in the terminal.

Kind regards,
Theodore.

**Comment by mimarilker** [comment]
I have updeted snappyHexMeshDict and SnappyHexMesh (beforehand, it couldn't accept data from snappyHexMeshDict, due to missing definitions in the coding) components, now they are compatible with explicitFeatureSnap! But eMesh file should be created manually for now, however I am working on it, I hope at the end, all of the process will be automated. Now I am sending pull requests for them, they wont replace the older components, I have tagged them with "new" title.

**Comment by DambronOlivier** [comment]
hi @thinklikeanarchitect,

I have been testing the new components you uploaded with the outdoor 2D.
Could you please let me know if the meshing is correct? One thing I came accross was to make sure to set the locationinMesh point out of any geometry. 
![capturemesh](https://cloud.githubusercontent.com/assets/21184323/22345405/dddeba60-e3f7-11e6-9412-c8ce67e81c5f.PNG)

Here is the log in case you need to look at it.
[snappyHexMesh.zip](https://github.com/ladybug-analysis-tools/butterfly/files/733234/snappyHexMesh.zip)

This is a capture after 3600 timesteps.
![capture](https://cloud.githubusercontent.com/assets/21184323/22345490/3528c87e-e3f8-11e6-81d5-0c780c7a5422.PNG)

Many thanks.

**Comment by thinklikeanarchitect** [comment]
Hi @DambronOlivier,

Please confirm the followings:

I. Have you created  the eMesh file manually in the directory of "triSurface" folder of your case? 
II. Is the name of the eMesh file is correct? I am asking that bcs "building.eMesh" was an example string, if you try this components on the outdoor_airflow sample, it would be "outdoor_airflow.eMesh". I know its very basic but please check.

Moreover I am working to automate all the process as I said and I am very near to success, just need a very basic help (@Mostapha). Now I have created a new comp. that creates "surfaceFeatureExtractDict" in the right folder instead of user (I am now sending a pull request for it and the wind tunnel comp). So user just run surfaceFeatureExtract command in the "start_OF" terminal. As an additional note(can be used as instruction later):

I. Run the terminal as an administrator.
II. Write 'cd workingDir/butterfly/yourCaseFolderName
III. Lastly, just run the 'surfaceFeatureExtract' command.

**Comment by thinklikeanarchitect** [comment]
Newer version of the "create case from wind tunnel" can be reached from the pull request that I sent before. The last one named "Add files via upload".

**Comment by thinklikeanarchitect** [comment]
I have improved these components and will share the updated ones soon!

**Comment by thinklikeanarchitect** [comment]
Now, explicit meshing works like a charm, without doing anything manually! I have been waiting for @mostaphaRoudsari to upload the components and python scripts.

**Comment by thinklikeanarchitect** [comment]
Just a sample from new components...
![logo](https://cloud.githubusercontent.com/assets/22752273/22399483/0e3dd484-e5a6-11e6-8781-72bc4f6df79d.JPG)

**Comment by mostaphaRoudsari** [comment]
This is now implemented in butterfly via b3c34ada6c1549593b033c3d646479d00f53bfc2. The implementation is simple. If the user set a value for `_surfaceFeatureLevel_` in snappyHexMeshDict then the
component will generate the `surfaceFeatureExtractDict` and runs the command before running `snappyHexMeshDict`. Let us know if there was any issues. 

![image](https://cloud.githubusercontent.com/assets/2915573/22481916/f2688706-e7c4-11e6-86ab-1b3c2836a97a.png)
